### DISCUSSION TITLE
Discussion for PopUp on iOS

### DISCUSSION THREAD

**Discussion by maonaoda** [CommunityToolkit/Maui#1461] [General]
@cat0363 
Sorry for @ you.

Is it possible to make MauiPopup`s ViewController set to public  for more freedom of customization.

such as 
1. I want to Add customization OverlayView for Popup.
2. Determine whether the current MauiPopUp ViewController is the previous Maui PopUp that has not been completely closed.
and reset the ViewController with wanted.
※I implemented LoadingIndicatorPage using PopUp, and after calling to close this Page, then show another PopUp. 
But at this time, there is a high probability that WindowStateManager.Default.GetCurrentUIViewController() will return previous MauiPopUp.

I am using ↓ to get (set) MauiPopup`s Internal (UIViewController)? named ViewController  currently.
But I don't want to use reflection if possible
```
    public class CustomMauiPopup(IMauiContext mauiContext) : MauiPopup(mauiContext)
    {
        public UIViewController? UIViewController => (UIViewController?)InternalAccessHelper.GetInternalProperty(this, 
"ViewController", true);
       ....
    }
```

(MauiPopup.macios.cs)
```
internal UIViewController? ViewController { get; private set; }
```

**Comment by cat0363** [comment]
@maonaoda , Since I don't have the authority to decide, I will cast one vote in favor of this discussion.
If the ViewController is exposed, the range of customization of MauiPopup will expand. I also agree with your opinion.

I think the discussion would be more lively if more people agreed, so if you agree with @maonaoda  opinion, please vote yes.

**Reply by cat0363** [reply]
@maonaoda , I created a PR for Issue #1452 and mentioned this Discussion there.

**Reply by maonaoda** [reply]
@cat0363 The pull request is very detailed. ありがとうございました～

**Comment by maonaoda** [comment]
For Android, it would be easy to customize the Dialog's DimBackgroundColor by adding another constructor with themeResId for MauiPopup.android.cs.
```
	public MauiPopup(Context context, IMauiContext mauiContext, int themeResId)
	: base(context, themeResId)
	{
		this.mauiContext = mauiContext ?? throw new ArgumentNullException(nameof(mauiContext));
	}
```

**Comment by CliffAgius** [comment]
Closed as per the Community Stand-up discussion here: https://youtu.be/N9wMcBP4jtg?t=2889

**Comment by tomShoutTheSecond** [comment]
> Is it possible to make MauiPopup`s ViewController set to public for more freedom of customization.
> 
> such as
> 
> 1. I want to Add customization OverlayView for Popup.
> 2. Determine whether the current MauiPopUp ViewController is the previous Maui PopUp that has not been completely closed.
>    and reset the ViewController with wanted.
>    ※I implemented LoadingIndicatorPage using PopUp, and after calling to close this Page, then show another PopUp.
>    But at this time, there is a high probability that WindowStateManager.Default.GetCurrentUIViewController() will return previous MauiPopUp.
> 
> I am using ↓ to get (set) MauiPopup`s Internal (UIViewController)? named ViewController currently. But I don't want to use reflection if possible
> 
> ```
>     public class CustomMauiPopup(IMauiContext mauiContext) : MauiPopup(mauiContext)
>     {
>         public UIViewController? UIViewController => (UIViewController?)InternalAccessHelper.GetInternalProperty(this, 
> "ViewController", true);
>        ....
>     }
> ```
> 
> (MauiPopup.macios.cs)
> 
> ```
> internal UIViewController? ViewController { get; private set; }
> ```

Hello could you please share your code for InternalAccessHelper? I am trying to implement your workaround for MAUI Editor alignment issues on iOS https://github.com/dotnet/maui/issues/18282#issuecomment-1778883149

Thank you!

**Reply by maonaoda** [reply]
@tomShoutTheSecond 
It seems had been fixed on https://github.com/dotnet/maui/pull/18672, OMG,Maui 8.0.3 is not having it yet?

```
        public static object GetInternalProperty(object target, string propertyName, bool baseType = false, BindingFlags bindingFlags = BindingFlags.Instance | BindingFlags.NonPublic)
        {
            var targetType = baseType ? target.GetType().BaseType : target.GetType();
            var propertyInfo = targetType.GetProperty(propertyName, bindingFlags);

            if (propertyInfo != null)
            {
                return propertyInfo.GetValue(target);
            }
            else
            {
                throw new ArgumentException($"Property '{propertyName}' not found on type '{targetType.FullName}'.");
            }
        }
```

**Comment by tomShoutTheSecond** [comment]
Yeah I'm still seeing the issue with the latest version of MAUI.
Unfortunately your fix isn't working for me either and the placeholder
doesn't show when I use your handler.

**Comment by maonaoda** [comment]
If you provide your handler, I might be able to help you take a look.

**Comment by tomShoutTheSecond** [comment]
Thanks, I have fixed the problem using my own handler:

```

	public partial class EditorHandler : ViewHandler<Editor, UITextView>
    {
        private UILabel _placeholder;

        protected override UITextView CreatePlatformView()
        {
            var platformView = new UITextView { Font = UIFont.SystemFontOfSize(14) };

            _placeholder = new UILabel
            {
                Text = "Enter text...",
                TextColor = ThemeColors.MidGrey.ToPlatform(),
                Font = UIFont.SystemFontOfSize(14),
                Frame = new CoreGraphics.CGRect(5, 0, 800, 32),
                ClipsToBounds = true
            };

            platformView.AddSubview(_placeholder);

            platformView.Changed += TextChanged;

            return platformView;
        }

        private void TextChanged(object sender, EventArgs e)
        {
            var textView = (UITextView)sender;

            _placeholder.Hidden = !textView.Text.IsNullOrEmpty();
        }

        protected override void DisconnectHandler(UITextView platformView)
        {
            base.DisconnectHandler(platformView);

            platformView.Changed -= TextChanged;
        }
    }

```
